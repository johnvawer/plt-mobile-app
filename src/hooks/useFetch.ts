import { useEffect, useState } from 'react'
import axios, { AxiosResponse } from 'axios'

interface ReturnPayload {
  data: AxiosResponse<any> | null
  loading: boolean
  error: boolean
}

const baseUrl = 'https://my-json-server.typicode.com/benirvingplt'

export const useFetch = (url: string): ReturnPayload => {
  const [loading, setLoading] = useState<boolean>(true)
  const [data, setData] = useState<AxiosResponse<any> | null>(null)
  const [error, setError] = useState<boolean>(false)

  useEffect(() => {
    if (url === '') return

    const fetchData = (): void => {
      setLoading(true)
      axios.get(`${baseUrl}${url}`)
        .then((response: AxiosResponse<any>) => {
          setData(response)
        })
        .catch(e => {
          console.log(e)
          setError(true)
        })
        .finally(() => {
          setLoading(false)
        })
    }

    fetchData()
  }, [url])

  return { loading, data, error }
}
