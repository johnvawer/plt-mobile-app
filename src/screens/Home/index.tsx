import React from 'react'
import { FlatList, Text } from 'react-native'

import Loading from '../../components/Atoms/Loading'
import ProductCard from '../../components/Molecules/ProductCard'
import { useFetch } from '../../hooks/useFetch'

interface Props {
  navigation: any
}

const Home = ({ navigation }: Props): JSX.Element => {
  const { loading, data, error } = useFetch('/products/products')

  if (loading) return <Loading />

  if (error || data === null) return <Text>Error Fetching data</Text>

  const products: Product[] = data.data

  return (
    <FlatList
      data={products}
      renderItem={({ item }: { item: Product }) => {
        const { id } = item

        const onPress = (id: number): void => {
          navigation.navigate('Product', {
            productId: id
          })
        }

        return (
          <ProductCard
            product={item}
            onPress={() => onPress(id)}
            buttonText='View Product'
          />
        )
      }}
    />
  )
}

export default Home
