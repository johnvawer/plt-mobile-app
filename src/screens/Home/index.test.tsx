import React from 'react'
import { render, fireEvent } from '@testing-library/react-native'
import * as fetchHook from '../../hooks/useFetch'
import Home from '.'

const fakeProduct = {
  id: 1,
  name: 'product name',
  colour: 'black',
  price: 10,
  img: 'http://someimage.png'
}

describe('<Home />', () => {
  let useFetchMock: jest.SpyInstance

  beforeEach(() => {
    useFetchMock = jest.spyOn(fetchHook, 'useFetch')
  })

  afterEach(() => {
    useFetchMock.mockRestore()
  })

  it('should call the api to fetch products', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: {
        data: [fakeProduct]
      },
      error: false
    })
    render(<Home navigation={{ navigate: jest.fn() }} />)

    expect(useFetchMock).toHaveBeenCalledTimes(1)
    expect(useFetchMock).toHaveBeenCalledWith('/products/products')
  })

  it('should display a loading message when the data is loading', () => {
    useFetchMock.mockReturnValueOnce({
      loading: true,
      data: null,
      error: false
    })
    const { toJSON } = render(<Home navigation={{ navigate: jest.fn() }} />)

    expect(toJSON()).toMatchSnapshot()
  })

  it('should display an error message if an error is returned', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: null,
      error: true
    })
    const { toJSON } = render(<Home navigation={{ navigate: jest.fn() }} />)

    expect(toJSON()).toMatchSnapshot()
  })

  it('should render the products correctly', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: {
        data: [fakeProduct]
      },
      error: true
    })
    const { toJSON } = render(<Home navigation={{ navigate: jest.fn() }} />)

    expect(toJSON()).toMatchSnapshot()
  })

  it('should navigate to the product screen on button press', () => {
    useFetchMock.mockReturnValueOnce({
      loading: false,
      data: {
        data: [fakeProduct]
      },
      error: false
    })

    const mockedNavigate = jest.fn()

    const { getByText } = render(<Home navigation={{ navigate: mockedNavigate }} />)
    const button = getByText('View Product')
    fireEvent.press(button)

    expect(mockedNavigate).toHaveBeenCalledTimes(1)
    expect(mockedNavigate).toHaveBeenCalledWith('Product', {
      productId: 1
    })
  })
})
