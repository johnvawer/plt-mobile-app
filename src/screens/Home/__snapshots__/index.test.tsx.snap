// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should display a loading message when the data is loading 1`] = `
<View>
  <Text>
    Loading
  </Text>
</View>
`;

exports[`<Home /> should display an error message if an error is returned 1`] = `
<Text>
  Error Fetching data
</Text>
`;

exports[`<Home /> should render the products correctly 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "colour": "black",
        "id": 1,
        "img": "http://someimage.png",
        "name": "product name",
        "price": 10,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        nativeID="animatedComponent"
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "margin": 10,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://someimage.png",
              }
            }
            style={
              Object {
                "height": 300,
                "marginBottom": 10,
                "width": 300,
              }
            }
          />
          <Text>
            product name
          </Text>
          <Text
            testID=""
          >
            Â£
            10.00
          </Text>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "borderColor": "black",
                "borderWidth": 1,
                "margin": 5,
                "padding": 10,
              }
            }
          >
            <Text>
              View Product
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
